Further refactoring

Rfp21 opened the pull request:
I have now:

- linked in the use of the game object to main. The game object creates the dictionary. The main programme calls the dictionary
- that being said, I think that either main and functions_for_main need to interact with Game a little more. There is some interaction in there but probably not fully complete
- Read through game - I have put a few comments / thoughts in
- Taken the enum from game and distrubuted it as comments in functions I think will need to use it
- Moved all of the game variables out of main and updated the list that was in game
- This means the the main parameters are now a list of inputs from the user as follows:
            - class load_game(Enum): 
                    NEW_GAME = 0
                    RESUME_GAME = 1
                    EXIT_GAME = 2
                    FACTORY_RESET = 3
            
            letter = None
- Written more verbose comments / ideas for Sam in Start_game function
- Suggested that word_selection is an import into Start_game, rather than into Game. I think that new game should call word_selection and game perhaps?
- Had a general tidy up / reduced problems
- I have totally re-written the workflow files from research. The test report is nearly working: Our badge on the README might update on this pull request
- I am still having difficulties with codecov. It recognises that recent pull requests have happenned but only has the main branch available. Rest of the workflow is working though I think.

Rfp21 committed:
Got rid of all chatgpt generated workflow - done research, trying again

Rfp21 committed:
think I got rid of directory error

Rfp21 committed:
more debug of new workflow

Rfp21 committed:
cleared data handling test to test pipeline

Rfp21 committed:
try again

Rfp21 committed:
try again

Rfp21 committed:
Fixed the architecture issues according to comments from merge 60

Rfp21 committed:
attempted to fix codecov again

Rfp21 committed:
try again

Rfp21 committed:
try again

Rfp21 committed:
try again

Rfp21 committed:
try again

Rfp21 committed:
try again

Rfp21 committed:
try again

Rfp21 committed:
tried to get black anfd flint to run every time in workflow

Rfp21 committed:
more workflow fixes - trying to get red and gren ticks

Rfp21 committed:
changed from steps to jobs - hoping to get green and red ticks

Rfp21 committed:
Think I have set up the jobs correctly in the workflows

Rfp21 committed:
fix to flake 8

Rfp21 committed:
hopefully last change today

Rfp21 committed:
trying to get codecov report

Rfp21 committed:
few more tweaks to main to avoid circular dependency

rickhardy commented:
# game_attributes -- None  of these can be in main, create an instance of game instead and make sure that the attributes are in there
guess_result = ["_","B","_","_","_","_"]
used_letters = ["_"]
attempts_remaining = 10
current_game_status = "Not Started"
guessed_word = [""]
words_guessed = [""]
current_score = 0
word = "" # will get word from word_selection
Start_Game_Selection = "" # will be click on New, Resume or Exit from the user,
game_index = ""
game_closed = False

Example of how to use the game class:

current_game = Game ()
print (current_game.score)
print (current_game.game_status)
current_game.game_status = current_game.Game_status (2)
print (current_game.game_status)
current_game.game_status = current_game.Game_status.LOST
print (current_game.game_status)

the function code needs to pass the current_game around and modify its attributes


# game_player attributes -- there needs to be another class called game player (for example), thats going to encapsulate all of the below. Its like the data constructor but more powerful
cumulative_score = 0
number_of_games_played = 0
number_of_games_won = 0
#-----------------------------------------------------------------------------
"""This is the constructor for the dictionary for the game data. It contains
lots of useful inputs to be supplied by the user from app.py or
from word_selection"""
#-----------------------------------------------------------------------------
data = {"word_progress": guess_result, 
        "used_letter": used_letters,
        "guessed_words": guessed_word,
        "game_status" : current_game_status,
        "remaining_attempts" : attempts_remaining,
        "score_keeping" : current_score,
        "cumulative_score" : cumulative_score,
        "number_games_played" : number_of_games_played,
        "number_games_won" : number_of_games_won,
        "current_game_index" : game_index,
        }

Rfp21 commented:
I'm a bit confused by the comments because I thought I had done that. 

I put this into data_handling

def initialise_game_and_data():
        data_constructor = Game()
        data = data_constructor.__dict__
        return data 

Then into main, I declared a variable called game which used the above function. Then I can use "game.attribute" in main,

I thought the above implements the comments you made / encapsulated but it could be me mis-understanding encapsulation

This is the version of main I am trying to pull:

# main program
#from word_selection import choose_word
#from game import Game
#from templates import Template

#-----------------------------------------------------------------------------
"""These are the functions used in the main programme"""
#-----------------------------------------------------------------------------

# these functions are all in the functions_for_main folder

#-----------------------------------------------------------------------------
from enum import Enum
from hangman_code.game import Game
from hangman_code.functions_for_main.validate_input import Validate_input
from hangman_code.functions_for_main.make_guess import Make_guess
from hangman_code.functions_for_main.game_status_function import Current_game_status
from hangman_code.functions_for_main.start_game import Start_game
from hangman_code.functions_for_main.game_status_function import setup_new_guess
from hangman_code.functions_for_main.game_status_function import is_won
from hangman_code.functions_for_main.game_status_function import is_lost
from hangman_code.functions_for_main.game_status_function import is_closed
from hangman_code.functions_for_main.data_handling import update_in_play_data
from hangman_code.functions_for_main.start_game import Start_game_Selection
from hangman_code.functions_for_main.data_handling import global_dictionary_update
#-----------------------------------------------------------------------------
"""This is me showing off my kudos"""
#-----------------------------------------------------------------------------
#JUST WANTED TO HIGHLIGHT THIS IS NOT CHATGPT GENERATED. PURE ROANNE
#-----------------------------------------------------------------------------
"""The list of public parameters for playing :
These are needed as user inputs from app"""
#-----------------------------------------------------------------------------
class load_game(Enum): 
        NEW_GAME = 0
        RESUME_GAME = 1
        EXIT_GAME = 2
        FACTORY_RESET = 3

letter = None

#-----------------------------------------------------------------------------
"""Below is the list of private parameters for running main"""
"""Else, the list of parameters for playing the Game is within Game.py"""
#-----------------------------------------------------------------------------

json_filename = "persistence.json"
game_closed = False
#-----------------------------------------------------------------------------
"""This is the constructor for the dictionary with various the 
inputs needed in functions_for_main.The initialised data itself is in 
game.py"""
#-----------------------------------------------------------------------------

# Now moved into data_handling so that all data and data handling are 
# kept together (Seperation of concerns)

#-----------------------------------------------------------------------------
"""These are the functions used in the main programme"""
#-----------------------------------------------------------------------------

# I have moved these functions into the functions_for_main folder

#-----------------------------------------------------------------------------
"""This is the actual game logic / flow / main programme"""
#-----------------------------------------------------------------------------
def play_game(load_game):
       
        Start_game_Selection(load_game)

        game = Start_game(letter, Start_game_Selection(load_game))
        # i.e. needs the letter from the user and the game object which is the
        #return from Start_game_selection

        while game.attempts_remaining > 0 & game_closed == False:

                try:
                       
                        Validate_input(letter, game.used_letters)

                except (TypeError, ValueError):

                        return "Code needs to be written to reset / try again"

                else:
                        results = Make_guess(letter, game.word, game.guess_result)
                        game.guess_result = results.get("word_progress")
                        current_game_status = Current_game_status(results)
                        #This will update the status of the game e.g.
                                # Is Won, Is Lost, In Play

                        if current_game_status == 1: #"In Play via enum"

                                in_play_game = update_in_play_data(
                                        game.used_letters, 
                                        game.guessed_word, 
                                        game.attempts_remaining, 
                                        game.current_score)
                                
                                setup_new_guess(in_play_game)
                                #This will re-set the screen to allow the user 
                                # to set up a new guess

                        elif current_game_status == 2: # "Is Won" in Enum
                               is_won()
                               Start_game_Selection(load_game)
                                # if new game, return to play game
                                #this logic should go into the
                                #Start_game_selection function
                                      
                               
                        elif current_game_status == 3: #"Is Lost" in Enum
                                is_lost()
                                Start_game_Selection(load_game)
                                # if new game, return to play game
                                #this logic should go into the
                                #Start_game_selection function

                finally:
                       is_closed(load_game == 4, json_filename, game)
                       #Logic to be worked out
         #The dictionary will now be updated - ready to store in persistence
         # if we wanted to resume the game later"""
                       return game_closed == True

rickhardy commented:
problem
this version gives an error when you press new game in the app

comment
main.py and  app.py have very similar functions. there should only be one of these,
play_game could be in the library files

Rfp21 commented:
My thinking is that we keep the main programme in main and pass signals to and from app.

Just to keep very obvious separation between what the user does and how they interact and what the programme does behind the scenes. 

I thought that app was the flask app so we would technically need app to interface with the user even if we moved 99% of logic into main?

I didn't realise we had library files. I'll take a look.

P.s, I've never tried clicking new game in flask or trying to link flask to main yet. I'll take a look

Rfp21 commented:
Ok, got it, think I understand what you are saying:

main - runs the programme : uses load_game / play_game etc
play_game - it is just another function - should be the interface with the rest of the hidden functions (in my mind)
app.py - Should be the interface with the user interactions (in my mind) and send signals to play_game

rickhardy commented:
Flask creates a session with a stack of parameters, current_game (an instance of game) should mirror the session. If the session gets updated by the user actions then current_game gets updated accordingly and if current_game is updated by the game logic then this is mirrored in the session. 

The session defines what is displayed in the html page

rickhardy commented:
This is what chat gpt says about the error 

Your traceback is telling you exactly what’s wrong:

re.sub('[a-zA-Z]', '_', word) is receiving a list, but re.sub only works on a string/bytes.

So somewhere, word is a list (e.g., ["apple"] or ["a", "p", "p", "l", "e"]) instead of "apple".

Where it blows up

hangman_code/game.py:

template = sub('[a-zA-Z]', '_', word)
# TypeError: expected string or bytes-like object, got 'list'

Why word became a list (most common causes)

You used random.sample(...) instead of random.choice(...)

random.sample(words, 1) returns a list of length 1 → ["apple"]

random.choice(words) returns a string → "apple"

You read words with something like .split() and then accidentally pass the whole list rather than one element.

Your “word source” function returns a list, but Game(...) expects a single word string.

Rfp21 commented:
Hmm.....I have never actually tried to pull a word as I was awaiting word_selection.

I'm just teaching myself html form actions and a little bit about flask and then I can revisit your comment with better understanding.

Rfp21 commented:
P.s, I think that the output from Johns word selection is a list

rickhardy commented:
This seems to work
